ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ExplicitDisclosureIT.scala

```diff
@@ -0,0 +1,421 @@
+      exerciseFetchError <- testContext
+        .exerciseFetchDelegated()
+        .mustFail("the submitter does not see the contract")
+
+      // Exercise the same choice, this time using correct explicit disclosure
+      _ <- testContext.exerciseFetchDelegated(testContext.disclosedContract)
+
+      // Create an extra disclosed contract
+      extraKey = ledger.nextKeyId()
+      _ <- ledger.create(owner, Delegated(owner, extraKey))
+      delegatedTxs <- ledger.flatTransactionsByTemplateId(Delegated.id, owner)
+      extraDelegatedEvent = createdEvents(delegatedTxs(1)).head
+      extraDisclosedContract = createEventToDisclosedContract(extraDelegatedEvent)
+
+      // Exercise the same choice, this time with superfluous disclosure
+      _ <- testContext.exerciseFetchDelegated(testContext.disclosedContract, extraDisclosedContract)
+    } yield {
+      assertGrpcError(
+        exerciseFetchError,
+        LedgerApiErrors.ConsistencyErrors.ContractNotFound,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
+    "EDExerciseByKeyDisclosedContract",
+    "A disclosed contract can be exercised by key by non-stakeholder if authorized",
+    partyAllocation = allocate(TwoParties),
+    enabled = _.explicitDisclosure,
+  ) { implicit ec =>
+    { case Participants(Participant(ledger, owner, divulgee)) =>
+      for {
+        // Create contract with `owner` as only stakeholder
+        _ <- ledger.create(owner, WithKey(owner))
+        withKeyTxIds <- ledger.flatTransactionsByTemplateId(WithKey.id, owner)
+        withKeyCreate = createdEvents(withKeyTxIds(1)).head
+        withKeyDisclosedContract = createEventToDisclosedContract(withKeyCreate)
+        // Assert that a random party can exercise the contract by key (if authorized)
+        // when passing the disclosed contract to the submission
+        _ <- ledger.submitAndWait(
+          exerciseWithKey_byKey_request(ledger, owner, divulgee, withKeyDisclosedContract)
+        )
+      } yield ()
+    }
+  }
+
+  test(
+    "EDMetadata",
+    "All create events have metadata defined",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    val contractKey = ledger.nextKeyId()
+    for {
+      _ <- ledger.create(owner, Delegated(owner, contractKey))
+      _ <- ledger.create(owner, Delegation(owner, delegate))
+      flats <- ledger.flatTransactions(owner)
+      trees <- ledger.transactionTrees(owner)
+      someTransactionId = flats.head.transactionId
+      flatById <- ledger.flatTransactionById(someTransactionId, owner)
+      treeById <- ledger.transactionTreeById(someTransactionId, owner)
+    } yield {
+      assertLength("flatTransactions", 2, flats)
+      assertLength("transactionTrees", 2, trees)
+      assert(
+        flats.map(createdEvents).forall(_.forall(_.metadata.isDefined)),
```

Comment:
should we also check what the metadata actually is? I suppose we can’t do much for the driver metadata but at least the other fields seem testable.